docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
+* `ClientDispatch`: Client dispatch table (see section 2.5 below).
+* `NpiSpecificCharacteristics`: Obtained from `ClientRegistrationInstance` parameter. This contains attach-type specific data that may be used by an extension for attaching an eBPF program. For example, when an eBPF program is being attached to an XDP hook, the network interface Id can be passed as attach parameter via this parameter. This tells the extension to invoke the eBPF program whenever there are any inbound packets on that network interface.
+
+The per-client data structure should be returned as the `ProviderBindingContext` output parameter.
+
+Upon [client detach callback](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/nc-netioddk-npi_provider_detach_client_fn) the provider must free the per-client context passed in via `ProviderBindingContext` parameter.
+
+### 2.5 Invoking an eBPF program from Hook NPI Provider
+To invoke an eBPF program, the extension uses the dispatch table supplied by the hook NPI client during attaching. There is only one function in the client dispatch table, that is of the following type:
+
+```
+/**
+ *  @brief This is the only function in the eBPF hook NPI client dispatch table.
+ */
+typedef ebpf_result_t (*ebpf_invoke_program_function_t)(
+    _In_ const void* client_binding_context, _In_ const void* context, _Out_ uint32_t* result);
+
+```
+The function pointer can be obtained from the client dispatch table as follows:
+```
+invoke_program = (ebpf_invoke_program_function_t)client_dispatch_table->function[0];
+```
+When an extension invokes this function pointer, then the call flows through the eBPF Execution Context and eventually invokes the eBPF program.
+When invoking an eBPF program, the extension must supply the client binding context it obtained from the hook NPI client as the `client_binding_context` parameter. For the second parameter `context`, it must pass the program type specific context data structure. Note that the program information NPI provider supplies information the context descriptor using `ebpf_context_descriptor_t` type to the eBPF verifier and JIT-compiler via the NPI client hosted by Execution Context. The `result` output parameter holds the return value from the eBPF program post execution.
+
+### 3.3 Authoring Helper Functions
+An extension can provide implementation of helper functions that can be invoked by the eBPF programs. The helper functions can be of two types:
+1. Program-Type specific: These helper functions can only be invoked by eBPF programs of a given program type. Usually, an extension may provide implementations for hooks of certain program types and provide helper functions that are associated with those helper functions. The program information NPI provider must then provide the prototypes and addresses for those functions. For these type of helpers, the helper function Id must be greater that 65535 (0xFFFF) for program type specific helper functions.
```

Comment:
https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md#program-types shows helpers that are supported by multiple (but not all) program types.  E.g., `BPF_FUNC_skb_load_bytes()` is supported for `BPF_PROG_TYPE_SOCKET_FILTER`, `BPF_PROG_TYPE_SCHED_CLS`, `BPF_PROG_TYPE_SCHED_ACT`, `BPF_PROG_TYPE_CGROUP_SKB`, `BPF_PROG_TYPE_SK_SKB`, `BPF_PROG_TYPE_SK_REUSEPORT`, and `BPF_PROG_TYPE_FLOW_DISSECTOR`.  How would we do that on Windows?

Reply:
This is simple if all the program types are in the same extension. The extension defines all the helpers. And then add the prototypes and addresses of the common helper functions to the program information data of each of these program times via the Program Info NPI provider.
If the program and helper function are in different drivers, then I suppose the driver implementing the helper functions also have to export them, such that the program info NPI provider on the other driver can refer to those functions.

Reply:
That doesn't allow someone to add helper functions to one of our program types via their own extension, which I think we will likely need in the future.  Ok to punt and leave it as an open issue, but maybe call it out as an open issue.

Reply:
Yes, I will start a GitHub discussion on this topic.